lib/grape/cookies.rb

```diff
@@ -2,41 +2,51 @@
 
 module Grape
   class Cookies
-    def initialize
-      @cookies = {}
-      @send_cookies = {}
-    end
+    extend Forwardable
 
-    def read(request)
-      request.cookies.each do |name, value|
-        @cookies[name.to_s] = value
-      end
+    DELETED_COOKIES_ATTRS = {
+      max_age: '0',
+      value: '',
+      expires: Time.at(0)
+    }.freeze
+
+    def_delegators :cookies, :[], :each
+
+    def initialize(lazy_rack_cookies)
+      @lazy_cookies = lazy_rack_cookies
     end
 
-    def write(header)
-      @cookies.select { |key, _value| @send_cookies[key] == true }.each do |name, value|
-        cookie_value = value.is_a?(Hash) ? value : { value: value }
-        Rack::Utils.set_cookie_header! header, name, cookie_value
+    def each_response_cookies
+      return unless defined?(@send_cookies)
+
+      send_cookies.each do |name|
+        yield name, cookies[name]
       end
     end
 
-    def [](name)
-      @cookies[name.to_s]
+    def []=(name, value)
+      cookies[name] = value
+      send_cookies << name
     end
 
-    def []=(name, value)
-      @cookies[name.to_s] = value
-      @send_cookies[name.to_s] = true
+    # see https://github.com/rack/rack/blob/main/lib/rack/utils.rb#L338-L340
+    def delete(name, **opts)
+      self.[]=(name, opts.merge(DELETED_COOKIES_ATTRS))
     end
 
-    def each(&block)
-      @cookies.each(&block)
+    private
+
+    def cookies
+      return @cookies if defined?(@cookies)
```

Comment:
The lazy cookies thing is an interesting way to solve this. I am not opposed to it, but before we merge, isn't it easier to set `@cookies = cookies` in `initialize` and `call` here? Avoids having two variables (less state), is easier to follow IMO.

```ruby
@cookies = @cookies.call.with_indifferent_access if @cookies.respond_to?(:call)
```